agent/api/task/task.go

```diff
@@ -3541,14 +3545,53 @@ func (task *Task) PopulateServiceConnectContainerMappingEnvVar() error {
 		if err != nil {
 			return fmt.Errorf("error retrieving task container for pause container %s: %+v", c.Name, err)
 		}
-		containerMapping[taskContainer.Name] = c.GetNetworkSettings().IPAddress
+		if instanceIPCompatibility.IsIPv6Only() && c.GetNetworkSettings().GlobalIPv6Address != "" {
+			containerMapping[taskContainer.Name] = c.GetNetworkSettings().GlobalIPv6Address
+		} else {
+			containerMapping[taskContainer.Name] = c.GetNetworkSettings().IPAddress
+		}
+	}
+	return task.setContainerMappingForServiceConnectContainer(containerMapping)
+}
+
+// PopulateServiceConnectContainerMappingEnvVarAwsvpc populates APPNET_CONTAINER_IP_MAPPING
+// environment variable for the AppNet Agent container (aka Service Connect container) for
+// awsvpc network mode tasks.
+// Application containers are expected to be bound to loopback address(es).
+func (task *Task) PopulateServiceConnectContainerMappingEnvVarAwsvpc() error {
+	primaryENI := task.GetPrimaryENI()
+	if primaryENI == nil {
+		return errors.New("no primary ENI found in task")
+	}
+	if !primaryENI.IsIPv6Only() {
+		// Explicit container address mapping is needed only for IPv6-only case.
+		// Service Connect sidecar defaults to IPv4 loopback in other cases.
+		return nil
+	}
+
+	containerMapping := make(map[string]string)
+	for _, c := range task.Containers {
+		if c == task.GetServiceConnectContainer() || c.IsInternal() {
```

Comment:
Just to verify, is this supposed to skip the container `GetServiceConnectContainer()` returns true for?

I am thinking it can be the side car container which we want to inject the mapping in, so the logic seems contradictory

Reply:
We want to add an IP mapping for each task container that was in the original task definition. So, the mapping doesn't need to include the sidecar container. The mapping will be provided to the sidecar container as an environment variable.

Reply:
Got it, thanks! I believe we did not skip the sidecar container when creating mapping in PopulateServiceConnectContainerMappingEnvVarBridge , so I might have been confused. Also seeing there is some update to this logic in latest commit which is skipping "internal" containers, which should be okay.

Reply:
Will this change - https://github.com/aws/amazon-ecs-agent/pull/4584/commits/2f913fff851dc99cbd722a9d76ffb403535a7244 be needed then?

Reply:
Confirmed that https://github.com/aws/amazon-ecs-agent/commit/2f913fff851dc99cbd722a9d76ffb403535a7244 change restores parity between bridge (logic for mapping already in place) and awsvpc network mode, so it is better to have same logic.